src/Qir/Runtime/stdlib/src/callables.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#![deny(clippy::all, clippy::pedantic)]
+
+use crate::{
+    arrays::{
+        QirArray, __quantum__rt__array_concatenate, __quantum__rt__array_update_reference_count,
+    },
+    tuples::{__quantum__rt__tuple_copy, __quantum__rt__tuple_update_reference_count},
+    update_counts,
+};
+use std::{cell::RefCell, rc::Rc, usize};
+
+#[derive(Clone)]
+pub struct Callable {
+    func_table: *mut *mut u8,
+    mem_table: *mut *mut u8,
+    cap_tuple: *mut u8,
+    is_adj: RefCell<bool>,
+    is_ctl: RefCell<u32>,
+}
+
+#[no_mangle]
+pub extern "C" fn __quantum__rt__callable_create(
+    func_table: *mut *mut u8,
+    mem_table: *mut *mut u8,
+    cap_tuple: *mut u8,
+) -> *const Callable {
+    Rc::into_raw(Rc::new(Callable {
+        func_table,
+        mem_table,
+        cap_tuple,
+        is_adj: RefCell::new(false),
+        is_ctl: RefCell::new(0),
+    }))
+}
+
+/// # Safety
+///
+/// This function should only be called with a callable created by `__quantum__rt__callable_*` functions.
+#[no_mangle]
+#[allow(clippy::cast_ptr_alignment)]
+pub unsafe extern "C" fn __quantum__rt__callable_invoke(
+    callable: *const Callable,
+    args_tup: *mut u8,
+    res_tup: *mut u8,
+) {
+    let call = Rc::from_raw(callable);
+    let index = (if *call.is_adj.borrow() { 1 } else { 0 })
+        + (if *call.is_ctl.borrow() > 0 { 2 } else { 0 });
+
+    // Collect any nested controls into a single control list.
+    let mut args_copy: *mut *const Vec<u8> = std::ptr::null_mut();
+    if !args_tup.is_null() {
+        // Copy the tuple so we can potentially edit it.
+        args_copy = __quantum__rt__tuple_copy(args_tup.cast::<*const Vec<u8>>(), true);
+
+        if *call.is_ctl.borrow() > 0 {
+            // If there are any controls, increment the reference count on the control list. This is just
+            // to balance the decrement that will happen in the loop and at the end of invoking the callable
+            // to ensure the original, non-owned list does not get incorrectly cleaned up.
+            __quantum__rt__array_update_reference_count(*args_copy.cast::<*const QirArray>(), 1);
+
+            let mut ctl_count = *call.is_ctl.borrow();
+            while ctl_count > 1 {
+                let ctls = *args_copy.cast::<*const QirArray>();
+                let inner_tuple = *args_copy
+                    .cast::<*const QirArray>()
+                    .wrapping_add(1)
+                    .cast::<*mut *const Vec<u8>>();
+                let inner_ctls = *inner_tuple.cast::<*const QirArray>();
```

Comment:
Can some of this code be abstracted into `unsafe` methods on `QirArray`?

Reply:
Perhaps, though I think the casts are relatively small and are the right way to go between different pointers. I could imagine that making sense in the future though.